dbms/src/Storages/MergeTree/MergeTreeSequentialBlockInputStream.cpp

```diff
@@ -0,0 +1,137 @@
+#include <Storages/MergeTree/MergeTreeSequentialBlockInputStream.h>
+#include <Storages/MergeTree/MergeTreeBlockReadUtils.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int MEMORY_LIMIT_EXCEEDED;
+}
+
+MergeTreeSequentialBlockInputStream::MergeTreeSequentialBlockInputStream(
+    const MergeTreeData & storage_,
+    const MergeTreeData::DataPartPtr & data_part_,
+    Names columns_to_read_,
+    bool read_with_direct_io_,
+    bool take_column_types_from_storage,
+    bool quiet)
+    : storage(storage_)
+    , data_part(data_part_)
+    , part_columns_lock(data_part->columns_lock)
+    , columns_to_read(columns_to_read_)
+    , read_with_direct_io(read_with_direct_io_)
+    , mark_cache(storage.context.getMarkCache())
+{
+    if (!quiet)
+        LOG_TRACE(log, "Reading " << data_part->marks_count << " marks from part " << data_part->name
+            << ", totaly " << data_part->rows_count
+            << " rows starting from the beginning of the part");
+
+    addTotalRowsApprox(data_part->rows_count);
+
+    header = storage.getSampleBlockForColumns(columns_to_read);
+    fixHeader(header);
+
+    /// Add columns because we don't want to read empty blocks
+    injectRequiredColumns(storage, data_part, columns_to_read);
```

Comment:
We break invariant `blocks are same with header`. All MergeTree*InputStreams break this invariant.